### DISCUSSION TITLE
awscli-like options that are shared across subcommands

### DISCUSSION THREAD

**Discussion by seanmcl** [bkirwi/decline#442] [Q&A]
The aws cli lets you do stuff like

```
aws --profile foo --region us-west-2 s3api get-bucket-policy --bucket foo
```

where there are 200 subcommands, but `--profile` and `--region` are shared across all of them. I know I can get close to this by creating opts for profile and region and just share them across all my subcommands.  I don't love this solution though. Apart from all the code copying, they only show up in the help for the subcommand, but not the help for the top command. I was hoping there's a way to structure the cli like

```
foo
  options for all subcommands of foo
  bar1
    options for bar1
  bar2
    options for bar2
  ...
```

Is this kind of thing possible today in Decline?

**Comment by bkirwi** [comment]
It is! I'm not able to respond in much detail now, but effectively doing something like `(profileOpt, regionOpt, subcommand1 orElse subcommand2).mapN(...)` works and has the behaviour you desire.

**Comment by bkirwi** [comment]
To elaborate a bit more: `Opts.subcommand` is an `Opts` instance like any other, and can be combined in all the usual ways. It's often used with `orElse`, but using `product`/`mapN` and friends works just as well when you want a subcommand _and_ some other options.

One final wrinkle to know about is that if you're using the ["effect style"](https://ben.kirw.in/decline/structure.html#config-and-effect-style), you'll generally end up with all your subcommands having a function type. One pattern some folks like is to come up with a special subcommand trait:

```scala
trait Subcommand {
  def run(profile: Profile, ...): IO[Unit]
}

val fooCommand: Opts[Subcommand] = ???
val profileOpt: Opts[Profile] = ???
(profileOpt, fooCommand).mapN((profile, sub) => sub.run(profile))
```

**Reply by seanmcl** [reply]
Thank you so much for your reply. Appreciate your time!
